presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayHashCodeOperator.java

```diff
@@ -45,16 +48,16 @@ private ArrayHashCodeOperator()
     public ScalarFunctionImplementation specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
     {
         Type type = types.get("T");
-        return new ScalarFunctionImplementation(false, ImmutableList.of(false), METHOD_HANDLE.bindTo(type), isDeterministic());
+        MethodHandle hashCodeFunction = functionRegistry.getScalarFunctionImplementation(internalOperator(HASH_CODE, BIGINT, ImmutableList.of(type))).getMethodHandle();
+        return new ScalarFunctionImplementation(false, ImmutableList.of(false), METHOD_HANDLE.bindTo(hashCodeFunction).bindTo(type), isDeterministic());
     }
 
-    public static long hash(Type type, Block block)
+    public static long hash(MethodHandle hashCodeFunction, Type type, Block block)
     {
-        int hash = 0;
+        long hash = 0;
         for (int i = 0; i < block.getPositionCount(); i++) {
             checkElementNotNull(block.isNull(i), ARRAY_NULL_ELEMENT_MSG);
-            // TODO: This should be migrated away from Type.hash and invoke a MethodHandle for the hash code operator instead
-            hash = (int) CombineHashFunction.getHash(hash, type.hash(block, i));
+            hash += hashPosition(hashCodeFunction, type, block, i);
```

Comment:
It needs to behave exactly the same as `ArrayType.hash`, so ya we should keep using `CombineHashFunction` and an `int`